SAK-48262 Portal spinner.js is not working properly

kunaljaykam opened the pull request:
https://sakaiproject.atlassian.net/browse/SAK-48262


Before:
![Screenshot from 2023-01-17 19-32-19](https://user-images.githubusercontent.com/50500283/212918806-5b98d726-8980-4db2-83dd-8d37f9d0db68.png)


After: 
![Screenshot from 2023-01-17 19-30-30](https://user-images.githubusercontent.com/50500283/212918853-af461312-9347-4844-8b9a-834fc9eef17d.png)

bjones86 commented on library/src/webapp/js/spinner.js:
```diff
@@ -118,7 +118,10 @@ SPNR.disableControlsAndSpin = function( clickedElement, escapeList )
+        escapeList = [
+            "current-site-button",
+            "sidebar-collapse-button",
+        ];
```
I think we should also add these defaults if `escapeList` is not null and doesn't already contain these.

kunaljaykam replied:
I also felt weird doing it, but nothing else came to mind. Please explain more about what I should do here.

bjones86 replied:
Maybe something like this:

```suggestion
    if( escapeList === null )
    {
        escapeList = [];
        escapeList = [
            "current-site-button",
            "sidebar-collapse-button"
        ];
    }
    else
    {
        if( !escapeList.includes("current-site-button") )
        {
            escapeList.push( "current-site-button" );
        }
        if( !escapeList.includes("sidebar-collapse-button") )
        {
            escapeList.push( "sidebar-collapse-button" );
        }
    }
```

I'm not very up-to-date on my JS, there may be a more elegant approach to do this.

adrianfish replied:
This could still result in multiple entries for the default strings. Maybe that doesn't matter though.

kunaljaykam replied:
escapeList = new Set(escapeList)
escapeList.add("current-site-button").add("sidebar-collapse-button")
escapeList = [...escapeList]


You are right, I think, I will commit @bjones86 suggestion. because it will be consistent with the rest of the code

bjones86 commented on library/src/skins/default/src/sass/base/_defaults.scss:
```diff
@@ -248,10 +248,11 @@ legend {
   box-shadow: var(--button-disabled-shadow);
   cursor: not-allowed;
   white-space: nowrap;
+  border-color: var(--button-disabled-border-color) !important;
```
I don't like seeing `!important`, is it absolutely necessary?

kunaljaykam replied:
It was necessary, because of b-5 btn class was overriding it

bjones86 commented on library/src/webapp/js/spinner.js:
```diff
@@ -353,7 +356,7 @@ SPNR.disableElementAndSpin = function( divID, element, activateSpinner )
         // <button type="button"> --> localName = "button", nodeName="button", tagName="BUTTON", type="button"
         element.tagName === "BUTTON" ? newElement.innerHTML = element.innerHTML : newElement.textContent = element.getAttribute( "value" );
-        newElement.className = activateSpinner ? "spinButton formButtonDisabled" : "formButtonDisabled";
+        newElement.className = element.className + " " + (activateSpinner ? "btn spinButton formButtonDisabled" : "btn formButtonDisabled");
```
I'm not sure why the "btn" class is necessary but I'm assuming you've determined it to be so.

kunaljaykam replied:
Without "btn", it won't clone the element with the same style and spinner

bjones86 replied:
Maybe something like this:

```suggestion
    if( escapeList === null )
    {
        escapeList = [];
        escapeList = [
            "current-site-button",
            "sidebar-collapse-button"
        ];
    }
    else
    {
        if( !escapeList.includes("current-site-button") )
        {
            escapeList.push( "current-site-button" );
        }
        if( !escapeList.includes("sidebar-collapse-button") )
        {
            escapeList.push( "sidebar-collapse-button" );
        }
    }
```

I'm not very up-to-date on my JS, there may be a more elegant approach to do this.

bjones86 reviewed (changes requested):
Small change requested, let's always add the defaults to the escape list whether it's null or not.

adrianfish commented:
No, don't commit Brian's change. Make it more compact. Just compact it down.

bjones86 commented:
I was thinking about this last night and remembered an idea I once had, but never had the time to implement.

Providing defaults for the escape list is good, but has some pitfalls: the IDs could change, or there could very well be new things included in the UI that need to be added to the escape list, and then all this has to be tracked down again.

Rather than taking this approach, what if we just confined the operation of spinner.js to a specific container. So for instance, we confine it to div class portal-main-container, or div ID content. In this way, the algorithms would never operate on items outside of the tool content, and we wouldn't have to worry about providing defaults for things that are in the outer portal.

It would be more changes, but I think it would be a lot more robust. We would have to change all the sub-routines to operate with query selectors, or similar, rather than just document.getElementsByTagName().

What do you guys think?

kunaljaykam commented:
IMO, It totally makes sense to implement what you suggest, I was thinking about the same but with just a different approach, like why not "remove spinner.js completely and then rewrite it to work when it's needed like a slow network is detected or server, then run a spinner or give more info("slow network, Sakai is getting ready"). because I think we don't really need spinner at all unless in some cases I think. I have actually tested the worksite setup without a spinner and it works just fine, you will not feel the need for a spinner unless your network or server is really slow. For reference read this article on spinner if you have time: 
https://uxdesign.cc/stop-using-a-loading-spinner-theres-something-better-d186194f771e
https://news.ycombinator.com/item?id=31231477


At last, I felt like more work for me, and we don't have time that much time right now, so I fixed it and moved on :)

kunaljaykam commented:
I understand, and If I am correct about your suggestion to contain the spinner.js inside the main tool container, I think it can be implemented, we can just tell it to get the elements from only inside id="content" / portal-main-container.  I will take another look at this. Thanks

adrianfish commented on library/src/webapp/js/spinner.js:
```diff
@@ -298,12 +298,16 @@ SPNR.disableTextAreas = function( escapeList )
  */
 SPNR.nodeListToArray = function( nodeList )
 {
```
return [ ...nodeList];

adrianfish replied:
Nah. Keep the function nodeListToArray, just make it one line in function body

bjones86 commented on library/src/webapp/js/spinner.js:
```diff
@@ -203,7 +206,7 @@ SPNR.disableInputs = function( clickedElement, escapeList )
 {
     // Get all the button elements
-    const buttons = SPNR.nodeListToArray( document.getElementsByTagName( "button" ) );
+    const buttons = content.document.getElementsByTagName( "button" );
```
I think this should just be content.getElementsByTagName rather than content.document.getElementsByTagName

```suggestion
    const buttons = content.getElementsByTagName( "button" );
```

bjones86 commented on library/src/webapp/js/spinner.js:
```diff
@@ -23,6 +23,9 @@
 var SPNR = SPNR || {};
 
+// Get the main content div
+const mainContent = document.getElementById( "content" ); 
```
Does this get updated with every request?

bjones86 commented:
Oh sorry, maybe I jumped the gun!

bjones86 reviewed (approved):
Nice! LGTM, thanks Kunal
